### PR TITLE
OKTA-701959 | [Dev Docs] issues with navigation on big pages

### PR DESCRIPTION
**Pull request by madhurishivakumar-okta**
## Description:
- **What's changed?** Remove caching and use binary search to get active anchors
- **Is this PR related to a Monolith release?** 

### Resolves:

* [OKTA-701959](https://oktainc.atlassian.net/browse/OKTA-701959)

### CODE REVIEW

File: packages/@okta/vuepress-theme-prose/mixins/AnchorHistory.vue [hunk 1]
```diff
@@ -67,23 +67,49 @@ export default {
...
+          offsetTop: parentHTag.offsetTop,
+          offsetHeight: parentHTag.offsetHeight
+        }
+      });
+
+      const scrollPosition = window.scrollY;
+      let start = 0;
+      let end = anchorOffsets.length - 1;
+
+      while (start <= end) {
```
Comment (emikhasyak-okta) [inline packages/@okta/vuepress-theme-prose/mixins/AnchorHistory.vue:95]: Here is my attempt:
```
function(anchors, scrollTo) {
      let end = anchors.length - 1;
      let start = 0;
      let mid = 0;
      let found = -1;

      while (start <= end) {
        mid = Math.floor((end + start) / 2);

        let offsetTop = anchors[mid].offsetTop;
        if (scrollTo == offsetTop) {
          return anchors[mid];
        }

        if (scrollTo > offsetTop) {
          start = mid + 1;
          found = mid;
        } else {
          end = mid - 1;
        }
      }
      return found !== -1 ? anchors[found] : 0;
 }
```
Your approach is easier to read, let's keep yours

**Review by emikhasyak-okta — COMMENTED** [review]
It's very close to the working solution. Could you refactor it a bit:
1. Obtaining scroll position. I don't know why it's max of those three scrolls and that's why I would leave it as it is
```   
const scrollTop = Math.max(
        window.pageYOffset,
        document.documentElement.scrollTop,
        document.body.scrollTop
     );
```
2. Instead of substructing `LAYOUT_CONSTANTS.ANCHOR_TOP_MARGIN` in every calculation, can we add it to scrollTop? I hope we won't need to re-adjust this with another constant
```   
const scrollTop = Math.max(
        window.pageYOffset,
        document.documentElement.scrollTop,
        document.body.scrollTop
     ) + LAYOUT_CONSTANTS.ANCHOR_TOP_MARGIN;
```
3. Remove `anchorOffsets` with the mapping and use `anchors` directly. So we have something like:
```
const scrollTop = ...

let start = 0;
let end = anchors.length - 1;
while (start <= end) {
   ...
}
```
Let's see if that works!

**Review by emikhasyak-okta — COMMENTED** [review]
We have two places where we need to initialize anchors array:
1. OnThisPage - we need anchors only when we scroll. I'd add initialization [here](https://github.com/okta/okta-developer-docs/blob/0aedd9e9db94b5e20cef92229097b2a32ad277e5/packages/%40okta/vuepress-theme-prose/components/OnThisPage.vue#L91). If this.anchors empty - this.anchors = this.getAnchors(). That way we will get array only once when it is needed.
2. ContentPage - it seems like we need an array only [here](https://github.com/okta/okta-developer-docs/blob/master/packages/%40okta/vuepress-theme-prose/components/ContentPage.vue#L79) and we can do the initialization once when anchors are empty.

What happens with component state when we navigate between pages `watch: { $page(to, from) { .. }}`? Let's say we stored anchors on page `docs.com/guideA` then navigated to `docs.com/guideB` - if anchors from `docs.com/guideA` are preserved we need to replace them with the new ones.

File: packages/@okta/vuepress-theme-prose/components/OnThisPage.vue [hunk 2]
```diff
@@ -107,19 +95,8 @@ export default {
     }, 200),
 
     updateAnchors: _.debounce(function () {
-      this.getAnchorsOffset();
       this.setActiveAnchor();
     }, 200),
-
-    getOnThisPageAnchors() {
```
[the remark this answers is not in the document]
Comment (madhurishivakumar-okta) [reply packages/@okta/vuepress-theme-prose/components/OnThisPage.vue:114]: I have moved the logic of getting the `on-this-page-link` anchors [here](https://github.com/okta/okta-developer-docs/pull/4711/files#diff-c44498602c0e96f34df301e59a8c9fe4f9d08bd264646fa6161fe6b3325f21f6R7-R12) in AnchorHistory.vue

[the remark this answers is not in the document]
Comment (emikhasyak-okta) [reply packages/@okta/vuepress-theme-prose/components/OnThisPage.vue:114]: A. In ContentPage we have all headers with `.header-anchor` class
B. In OnThisPage we have all header with `.header-anchor` class which **does not** have `.on-this-page-link` class
A != B it's different arrays. It's not possible to initialize anchors in AnchorHistory class and have both A and B set up correctly.
In the current implementation OnThisPage object will work as before and ContentPage object will work differently. 

After some digging, I think we can leave the current approach. In ContentPage we still ignore headers with `.on-this-page-link` class, it's confusing :(

**Review by emikhasyak-okta — APPROVED** [review]
Great job! 🚢

File: packages/@okta/vuepress-theme-prose/components/OnThisPage.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (madhurishivakumar-okta) [reply packages/@okta/vuepress-theme-prose/components/OnThisPage.vue:114]: I believe the final anchor array will consist of elements from the `.header-anchor` array whose hashes are also found in `.on-this-page-link`

**Pull request merged**